File: practice-dsa-with-vikas/src/com/vj/misc/oldCCodes/mergeSort.c
#include <stdio.h>
#include <stdlib.h>
int *b = NULL;

void merging(int *a, int low, int mid, int high)
{
    int l1, l2, i;
    for (l1 = low, l2 = mid + 1, i = low; l1 <= mid && l2 <= high; i++)
    {
        if (a[l1] <= a[l2])
            b[i] = a[l1++];
        else
            b[i] = a[l2++];
    }
    while (l1 <= mid)
        b[i++] = a[l1++];
    while (l2 <= high)
        b[i++] = a[l2++];
    for (i = low; i <= high; i++)
        a[i] = b[i];
}
void sort(int *a, int low, int high)
{
    int mid;
    if (low < high)
    {
        mid = (low + high) / 2;
        sort(a, low, mid);
        sort(a, mid + 1, high);
        merging(a, low, mid, high);
    }
    else
    {
        return;
    }
}

void display(int *arr, int size)
{
    printf("\n[ ");
    for (int i = 0; i < size; i++)
    {
        printf("%d, ", arr[i]);
    }
    printf("] ");
}

int main(int argc, char *argv[])
{
    int size = atoi(argv[1]);
    int *array = (int *)malloc(size * sizeof(int));
    b = (int *)malloc(size * sizeof(int));

    for (int i = 2, j = 0; i < size + 2; i++)
    {
        array[j++] = atoi(argv[i]);
    }
    printf("\n Before");
    display(array, size);
    sort(array, 0, size - 1);
    printf("\n After");
    display(b, size);
    return 0;
}

File: dsa-all-in-one-vikas/src/com/vj/searching_and_sorting/sorting/impl/MergeSort.java
package com.vj.searching_and_sorting.sorting.impl;

import java.util.*;

/** References **/
//1. https://takeuforward.org/data-structure/merge-sort-algorithm/

class Solution {

	private static void merge(int[] arr, int low, int mid, int high) {
		ArrayList<Integer> temp = new ArrayList<>(); // temporary array of size (high-low + 1)
		int left = low; // starting index of left half of arr
		int right = mid + 1; // starting index of right half of arr

		// storing elements in the temporary array in a sorted manner//

		while (left <= mid && right <= high) {
			if (arr[left] <= arr[right]) {
				temp.add(arr[left]);
				left++;
			} else {
				temp.add(arr[right]);
				right++;
			}
		}

		// if elements on the left half are still left //

		while (left <= mid) {
			temp.add(arr[left]);
			left++;
		}

		// if elements on the right half are still left //
		while (right <= high) {
			temp.add(arr[right]);
			right++;
		}

		// transfering all elements from temporary to arr //
		// hence merge sort is not-inplace: requires: auxillary array of size n

//		int z = 0;
		for (int i = low; i <= high; i++) {
			arr[i] = temp.get(i - low);
//			arr[i] = temp.get(z++);
		}
	}

	public static void mergeSort(int[] arr, int low, int high) {
		if (low >= high)
			return;
		int mid = (low + high) / 2;
		mergeSort(arr, low, mid); // left half
		mergeSort(arr, mid + 1, high); // right half
		merge(arr, low, mid, high); // merging sorted halves
	}
}

public class MergeSort {
	public static void main(String args[]) {
//		Scanner sc = new Scanner(System.in);

		int[] arr = { 4, 6, 2, 5, 7, 9, 1, 9, 0, 8, 6, 3 };
		int n = arr.length;
		System.out.println("Before sorting array: ");
		System.out.print(Arrays.toString(arr));
		System.out.println();

		Solution.mergeSort(arr, 0, n - 1);
		System.out.println("After sorting array: ");
		System.out.print(Arrays.toString(arr));

		System.out.println();
	}
}
